Refresh concepts/entity-template (#920)
spatialos/gdk-for-unity: docs/reference/concepts/entity-templates.md

<%(TOC)%>

# Entity templates

<%(Callout message="
Before reading this document, make sure you are familiar with:

  * [MonoBehaviour and ECS workflows]({{urlRoot}}/reference/workflows/which-workflow)
  * [Workers in the GDK]({{urlRoot}}/reference/concepts/worker)
  * [Standard schema library components](https://docs.improbable.io/reference/latest/shared/glossary#standard-schema-library-components)
")%>

> An [`EntityTemplate`]({{urlRoot}}/api/core/entity-template) specifies what [components]({{urlRoot}}/reference/glossary#spatialos-component) a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) will have, the initial values of these components and which [layers](https://docs.improbable.io/reference/latest/shared/glossary#layers) get [write access]({{urlRoot}}/reference//glossary#authority) to each component.

Reviewer: ```suggestion
An [`EntityTemplate`]({{urlRoot}}/api/core/entity-template) specifies what [components]({{urlRoot}}/reference/glossary#spatialos-component) a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) contains, the initial values of these components and which [layers](https://docs.improbable.io/reference/latest/shared/glossary#layers) get [write access]({{urlRoot}}/reference//glossary#authority) to each component.
```

## How to define an entity template

To define an entity template, you must first construct an `EntityTemplate`.

```csharp
var entityTemplate = new EntityTemplate();
```

An `EntityTemplate` can be mutated and used multiple times.

### Add components to entity template

All SpatialOS entities require the `Position` and `EntityAcl` components.

> The `Position` component must be added to the entity template manually. It is used by SpatialOS for load-balancing purposes and some types of queries.

Reviewer: do we want to be more specific about what kind of queries?

Author: Ah I forgot to link this to a couple of other docs pages

I think we could give an example e.g. relative constraints

> The `EntityAcl` component is automatically handled by the `EntityTemplate` class. It is this component that determines which types of workers have read access to an entity and, for each component, which type of worker can have write access.

Reviewer: ```suggestion
The `EntityAcl` component is automatically handled by the `EntityTemplate` class. This component determines which types of workers have read access to an entity and, for each component, which type of worker can have write access.
```

should we add a note that only one worker can be authoritative?

> > Understand more about the standard schema library components [here](https://docs.improbable.io/reference/latest/shared/glossary#standard-schema-library-components).

Reviewer: ```suggestion
> To learn more about the standard schema library components, [go to this documentation](https://docs.improbable.io/reference/latest/shared/glossary#standard-schema-library-components).
```

> To add components to the entity template, use the `AddComponent` method. For example, to add a `Position` component with [write access]({{urlRoot}}/reference//glossary#authority) given to UnityGameLogic workers:

Reviewer: ```suggestion
To add components to the entity template and specify which worker type has write access over the component, use the `AddComponent` method.
```

```csharp

> entityTemplate.AddComponent(new Position.Snapshot(coords), "UnityGameLogic");

Reviewer: ```suggestion
// Adds the Position component to the template and specifies that only a UnityGameLogic worker can have write access.
entityTemplate.AddComponent(new Position.Snapshot(coords), "UnityGameLogic");
```

Author: updated

<%(TOC)%>

# Entity templates

<%(Callout message="
Before reading this document, make sure you are familiar with:

  * [MonoBehaviour and ECS workflows]({{urlRoot}}/reference/workflows/which-workflow)
  * [Workers in the GDK]({{urlRoot}}/reference/concepts/worker)
  * [Standard schema library components](https://docs.improbable.io/reference/latest/shared/glossary#standard-schema-library-components)
")%>

An [`EntityTemplate`]({{urlRoot}}/api/core/entity-template) specifies what [components]({{urlRoot}}/reference/glossary#spatialos-component) a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) contains, the initial values of these components and which [layers](https://docs.improbable.io/reference/latest/shared/glossary#layers) get [write access]({{urlRoot}}/reference//glossary#authority) to each component.

## How to define an entity template

> To define an entity template, you must first construct an `EntityTemplate`.

Reviewer 2: O rly
